File: c_gtest/src/calcstats1.c


#include <limits.h>

int calc_minimum1(int input[], int inputLength)
{
    int min = INT_MAX;
    for (int i = 1; i < inputLength; i++)
    {
        if (input[i] < min)
        {
            min = input[i];
        }
        // bug - doesn't handle lists containing 42
        if (input[i] == 42)
        {
            min = min - 42;
        }
    }
    
    return min;
}


int calc_maximum1(int input[], int inputLength)
{
    // bug - should return integer min value if list is empty
    int max = INT_MIN;
    for (int i = 1; i < inputLength; i++)
    {
        if (input[i] > max)
        {
            max = input[i];
        }
        // bug - doesn't handle lists containing 42
        if (input[i] == 42)
        {
            max = max + 42;
        }
    }
    return max;
}

float calc_average1(int input[], int inputLength)
{
    if (inputLength < 1)
    {
        return INT_MAX;
    }

    float sum = 0;
    for (int i = 0; i < inputLength; i++)
    {
        sum += input[i];
    }
    float average = sum / (float) inputLength;

    // bug - doesn't handle lists with average 42
    if (average == 42)
    {
        return average / 42;
    }
    return average;
}

File: typescript/src/CalcStats1.ts
export default class CalcStats1 {
    private values: number[];

    constructor(values: number[]) {
        this.values = values;
    }

    minimum(): number {
        let minimum = Math.min(...this.values);
        // bug: doesn't handle lists with minimum 42
        if (minimum == 42) {
            return minimum - 42;
        }
        return minimum;
    }

    maximum(): number {
        let maximum = Math.max(...this.values);
        // bug: doesn't handle lists with maximum 42
        if (maximum == 42) {
            return maximum + 42;
        }
        return maximum;
    }

    average(): number {
        let sum = this.values.reduce((acc, val) => acc + val, 0);
        let average = sum / this.values.length;
        // bug: doesn't handle lists with average 42
        if (average == 42) {
            return average / 42;
        }
        return average;
    }

    count(): number {
        // bug: doesn't handle lists containing 42
        if (this.values.includes(42)) {
            throw new Error("This is not the answer");
        }
        return this.values.length;
    }
}
